File: testISA.m
% testISA computes p, rho, T, asound in ISA from SL to 31000 m

clc 
clear

toPlot = false; % set true for plotting

hmax = 31000; % m
h = [0:1000:hmax]./0.3042; % m -> ft
for i = 1:length(h)
    T(i) = temperature(h(i),'uk');
    p(i) = pressure(h(i),'uk');
    d(i) = density(h(i),'uk');
    a(i) = asound(h(i),'uk');
end

if toPlot
    % temperature
    figure 
    plot(T,h)
    xlabel('Temperature (K)')
    ylabel('h (ft)')
    % pressure
    figure 
    plot(p,h)
    xlabel('Pressure (psf)')
    ylabel('h (ft)')
    % density
    figure 
    plot(p,h)
    xlabel('Density (slig/ft^3)')
    ylabel('h (ft)')
    % asound
    figure 
    plot(a,h)
    xlabel('Sound speed (kt)')
    ylabel('h (ft))
end

% Check values with https://www.digitaldutch.com/atmoscalc/

hr = 5000; % ft
Tr = temperature(hr,'uk') % K
pr = pressure(hr,'uk') % psf
dr = density(hr,'uk') % slug/ft^3
ar = asound(hr,'uk') % kt
Tre = 278.244; % K
pre = 1760.80; % psf
dre = 0.00204834; % sluf/ft^3
are = 650.010; % kt
if abs(Tr-Tre)<=eps
    error('Temp @ 5000 wrong')
else
    disp('Temp @ 5000 ft OK')
end
if abs(pr-pre)<=eps
    error('Press @ 5000 wrong')
else
    disp('Press @ 5000 ft OK')
end
if abs(dr-dre)<=eps
    error('Dens @ 5000 wrong')
else
    disp('Dens @ 5000 ft OK')
end
if abs(ar-are)<=eps
    error('Asound @ 5000 wrong')
else
    disp('Asound @ 5000 ft OK')
end


hr = 25700; % m
Tr = temperature(hr) % K
pr = pressure(hr) % psf
dr = density(hr) % slug/ft^3
ar = asound(hr) % kt
Tre = 222.350; % K
pre = 2254.59; % psf
dre = 0.0353239; % sluf/ft^3
are = 298.926; % kt
if abs(Tr-Tre)<=eps
    error('Temp @ 25700 wrong')
else
    disp('Temp @ 25700 ft OK')
end
if abs(pr-pre)<=eps
    error('Press @ 25700 wrong')
else
    disp('Press @ 25700 ft OK')
end
if abs(dr-dre)<=eps
    error('Dens @ 25700 wrong')
else
    disp('Dens @ 25700 ft OK')
end
if abs(ar-are)<=eps
    error('Asound @ 25700 wrong')
else
    disp('Asound @ 25700 ft OK')
end

File: testISA.py
from isa.temperature import temperature 
from isa.pressure import pressure 
from isa.asound import asound 
from isa.density import density 
import numpy as np
import matplotlib.pyplot as plt


hmax = 31000  # m
h = np.linspace(0, hmax, int(hmax/1000))
T = np.zeros(len(h))
d = np.zeros(len(h))
p = np.zeros(len(h))
a = np.zeros(len(h))
for i in range(len(h)):
    T[i] = temperature(h[i])
    p[i] = pressure(h[i])
    d[i] = density(h[i])
    a[i] = asound(h[i])


fig, axs = plt.subplots(2, 2)
fig.suptitle('ISA Atmosphere')
axs[0, 0].plot(d, h)
axs[0, 0].set_xlabel('Density (kg/m^3)')
axs[0, 1].plot(p, h)
axs[0, 1].set_xlabel('Pressure (Pa)')
axs[1, 0].plot(a, h)
axs[1, 0].set_xlabel('Sound speed (m/s)')
axs[1, 1].plot(T, h)
axs[1, 1].set_xlabel('Temperature (K)')

plt.show()
